File: lib/core/models/patient.dart
// Patient model — represents the authenticated patient's profile
class Patient {
  final String id;
  final String name;
  final String phone;
  final int age;
  final String gender;
  final bool isPregnant;
  final int? gestationalWeek;
  final String? edd;
  final String village;
  final String district;
  final String state;
  final String bloodGroup;
  final String emergencyContactName;
  final String emergencyContactPhone;
  final String abhaId;
  final String preferredLanguage;
  final List<String> allergies;
  final List<String> conditions;

  const Patient({
    required this.id,
    required this.name,
    required this.phone,
    required this.age,
    required this.gender,
    this.isPregnant = false,
    this.gestationalWeek,
    this.edd,
    required this.village,
    required this.district,
    required this.state,
    required this.bloodGroup,
    this.emergencyContactName = '',
    this.emergencyContactPhone = '',
    this.abhaId = '',
    this.preferredLanguage = 'en',
    required this.allergies,
    required this.conditions,
  });

  factory Patient.fromJson(Map<String, dynamic> json) {
    return Patient(
      id: json['id'] as String? ?? '',
      name: json['name'] as String? ?? '',
      phone: json['phone'] as String? ?? '',
      age: (json['age'] as num?)?.toInt() ?? 0,
      gender: json['gender'] as String? ?? 'Female',
      isPregnant: (json['isPregnant'] ?? json['is_pregnant']) as bool? ?? false,
      gestationalWeek:
          ((json['gestationalWeek'] ?? json['gestational_week']) as num?)
              ?.toInt(),
      edd: (json['edd'] ?? json['estimated_due_date']) as String?,
      village: json['village'] as String? ?? '',
      district: json['district'] as String? ?? '',
      state: json['state'] as String? ?? 'Maharashtra',
      bloodGroup: json['bloodGroup'] as String? ?? '',
      emergencyContactName:
          (json['emergencyContactName'] ?? json['emergency_contact_name'])
                  as String? ??
              '',
      emergencyContactPhone:
          (json['emergencyContactPhone'] ?? json['emergency_contact_phone'])
                  as String? ??
              '',
      abhaId: (json['abhaId'] ?? json['abha_id']) as String? ?? '',
      preferredLanguage:
          (json['preferredLanguage'] ?? json['preferred_language'])
                  as String? ??
              'en',
      allergies:
          (json['allergies'] as List<dynamic>?)
              ?.map((e) => e.toString())
              .toList() ??
          const [],
      conditions:
          (json['conditions'] as List<dynamic>?)
              ?.map((e) => e.toString())
              .toList() ??
          const [],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'name': name,
      'phone': phone,
      'age': age,
      'gender': gender,
      'isPregnant': isPregnant,
      'gestationalWeek': gestationalWeek,
      'edd': edd,
      'village': village,
      'district': district,
      'state': state,
      'bloodGroup': bloodGroup,
      'emergencyContactName': emergencyContactName,
      'emergencyContactPhone': emergencyContactPhone,
      'abhaId': abhaId,
      'preferredLanguage': preferredLanguage,
      'allergies': allergies,
      'conditions': conditions,
    };
  }

  Patient copyWith({
    String? id,
    String? name,
    String? phone,
    int? age,
    String? gender,
    bool? isPregnant,
    int? gestationalWeek,
    String? edd,
    String? village,
    String? district,
    String? state,
    String? bloodGroup,
    String? emergencyContactName,
    String? emergencyContactPhone,
    String? abhaId,
    String? preferredLanguage,
    List<String>? allergies,
    List<String>? conditions,
  }) {
    return Patient(
      id: id ?? this.id,
      name: name ?? this.name,
      phone: phone ?? this.phone,
      age: age ?? this.age,
      gender: gender ?? this.gender,
      isPregnant: isPregnant ?? this.isPregnant,
      gestationalWeek: gestationalWeek ?? this.gestationalWeek,
      edd: edd ?? this.edd,
      village: village ?? this.village,
      district: district ?? this.district,
      state: state ?? this.state,
      bloodGroup: bloodGroup ?? this.bloodGroup,
      emergencyContactName: emergencyContactName ?? this.emergencyContactName,
      emergencyContactPhone: emergencyContactPhone ?? this.emergencyContactPhone,
      abhaId: abhaId ?? this.abhaId,
      preferredLanguage: preferredLanguage ?? this.preferredLanguage,
      allergies: allergies ?? this.allergies,
      conditions: conditions ?? this.conditions,
    );
  }
}

File: backend/src/models/Patient.js
// Patient model — associates a Firebase authenticated user with a patient
// profile. Identity (firebaseUid, phone) is derived ONLY from the verified
// Firebase ID token. Profile fields (Phase 3) are set/updated by the patient
// and are never trusted for identity lookups.
const mongoose = require('mongoose');

const patientSchema = new mongoose.Schema(
    {
        // Firebase UID — the authoritative identity key.
        // Derived from the verified Firebase ID token, NEVER from request bodies.
        firebaseUid: {
            type: String,
            required: true,
            unique: true,
            index: true,
        },

        // Phone number from the verified Firebase token (E.164 format, e.g. +919876543210)
        phone: {
            type: String,
            required: true,
        },

        // ── Patient profile (Phase 3) ──────────────────────────────────────
        // These are the patient's own editable profile fields. They are
        // stored on the patient record, keyed by firebaseUid. Defaults keep
        // existing patients (created before Phase 3) valid.
        fullName: {
            type: String,
            default: '',
            trim: true,
        },

        age: {
            type: Number,
            default: 0,
            min: 0,
            max: 150,
        },

        gender: {
            type: String,
            enum: ['Female', 'Male', 'Other', ''],
            default: '',
        },

        village: {
            type: String,
            default: '',
            trim: true,
        },

        district: {
            type: String,
            default: '',
            trim: true,
        },

        state: {
            type: String,
            default: '',
            trim: true,
        },

        bloodGroup: {
            type: String,
            enum: ['A+', 'A-', 'B+', 'B-', 'O+', 'O-', 'AB+', 'AB-', "Don't Know", ''],
            default: '',
        },

        allergies: {
            type: [String],
            default: [],
        },

        conditions: {
            type: [String],
            default: [],
        },

        // Role is fixed server-side. Clients cannot set or change it.
        role: {
            type: String,
            enum: ['PATIENT'],
            default: 'PATIENT',
        },

        isActive: {
            type: Boolean,
            default: true,
        },

        lastLoginAt: {
            type: Date,
            default: Date.now,
        },
    },
    {
        timestamps: true, // adds createdAt / updatedAt
        versionKey: false,
    }
);

module.exports = mongoose.model('Patient', patientSchema);